src/uu/dd/src/dd.rs

```diff
@@ -280,51 +280,164 @@ impl<'a, R: Read> Input<'a, R> {
     }
 }
 
-trait OutputTrait: Sized + Write {
-    fn fsync(&mut self) -> io::Result<()>;
-    fn fdatasync(&mut self) -> io::Result<()>;
+enum Density {
+    Sparse,
+    Dense,
 }
 
-struct Output<'a, W: Write> {
-    dst: W,
-    settings: &'a Settings,
+/// Data destinations.
+enum Dest {
+    /// Output to stdout.
+    Stdout(Stdout),
+
+    /// Output to a file.
+    ///
+    /// The [`Density`] component indicates whether to attempt to
+    /// write a sparse file when all-zero blocks are encountered.
+    File(File, Density),
 }
 
-impl<'a> Output<'a, io::Stdout> {
-    fn new(settings: &'a Settings) -> UResult<Self> {
-        let mut dst = io::stdout();
+impl Dest {
+    fn fsync(&mut self) -> io::Result<()> {
+        match self {
+            Self::Stdout(stdout) => stdout.flush(),
+            Self::File(f, _) => {
+                f.flush()?;
+                f.sync_all()
+            }
+        }
+    }
 
-        // stdout is not seekable, so we just write null bytes.
-        if settings.seek > 0 {
-            io::copy(&mut io::repeat(0u8).take(settings.seek as u64), &mut dst)
-                .map_err_context(|| String::from("write error"))?;
+    fn fdatasync(&mut self) -> io::Result<()> {
+        match self {
+            Self::Stdout(stdout) => stdout.flush(),
+            Self::File(f, _) => {
+                f.flush()?;
+                f.sync_data()
+            }
         }
+    }
 
-        Ok(Self { dst, settings })
+    fn seek(&mut self, n: u64) -> io::Result<u64> {
+        match self {
+            Self::Stdout(stdout) => {
+                let mut zeros = io::repeat(0u8).take(n as u64);
```

Comment:
One more nit, the `u8` and `as u64` here are unnecessary I believe. The same goes for the `u8` in `0u8` in `is_sparse`, though I understand if you want to keep these suffixes to make it clear that these are single bytes. `n` in this function is already a `u64` though, so the cast is not needed.

Reply:
Oops, those are leftovers from an earlier working version of the code. I'll fix that.